### ISSUE TITLE
VM too restrictive when analyzing potentially constant expressions

### ISSUE THREAD

**Issue by mhausner** [dart-lang/sdk#24970]
The VM fails to compile this code:

class A {
  final String s;
  const A(int i) : s = "$i";
}

main() {
  const A(88);
}

error: line 5 pos 20: initializer expression must be compile time constant.
  const A(int i) : s = "$i";

The initializer is a legal "potentially constant expression" as described in the language spec.

The analyzer and dart2js are doing the right thing.

**Comment by munificent** [comment]
@jtmcdole

**Comment by jtmcdole** [comment]
Woot

**Comment by lrhn** [comment]
@gbracha 

I'm not sure it should work.

The specification isn't absolutely clear on the restrictions on potentially constant expressions.
It just says that they would be compile-time constant expressions if parameters are treated as simple-typed constants. It should probably also say that the expression must also be a valid expression if the variables are non-constants.

I can't see text that differentiates between  `"${x}"` and `const C(x)` as potentially constant expressions - both productions requires the sub-expressions to be compile-time constants, but yet we allow the former and not the latter.

There are good reasons for not allowing the latter (`const C(x)` doesn't work when x is *not* a constant), but a *consistent* reading of the specification should allow either both or neither, since they are specified the same way.

If we add "and must be a valid expression when variables are treated as non-constant", then `"$x"` would be valid and `const C(x)`/`const [x]`/`const {x:x}` would not.

**Comment by iposva-google** [comment]
Thanks Lasse! Assigning to language until the ambiguity is resolved.

**Comment by mhausner** [comment]
If Gilad confirms that the code in this issue is legal, this CL will fix the bug:
https://codereview.chromium.org/1455213002/

**Comment by jtmcdole** [comment]
16.1 Constants 
A constant expression is an expression whose value can never change, and
that can be evaluated entirely at compile time.
A constant expression is one of the following:
• A literal number (16.3).
• A literal boolean (16.4).
• A literal string (16.5) where any interpolated expression (16.5.1) is a
compile-time constant that *evaluates to a numeric*, string or boolean value
or to null. It would be tempting to allow string interpolation where the interpolated
value is any compile-time constant. However, this would require
running the toString() method for constant objects, which could contain arbitrary
code

**Comment by mhausner** [comment]
Exactly. According to my understanding, the string interpolation in this example is a const expression when the constructor is called with a number or a string or a boolean as the actual parameter, but it would not be a constant if the constructor is called with, for example, const Point(1,1) or const [1,2,3] as the actual parameter.

**Comment by lrhn** [comment]
Allowing `"a${x}b"` when `x` is stringifyable is *probably* what we want, and I can't see a problem with having it (since other implementations already does so), but I don't think the specification is conclusive either way.

An expression `const {x:y}` is a compile-time constant if x and y are compile time constants and the class of x doesn't override Object.== - or it's int, string, etc.

Given that, the constructor `const C(x) : y = const {x: x};` satisfies the *stated* constraints of a const constructor just as well as `const C(x) : y = "${x}";` does. In practice, the former is not valid and the latter is. This boils down to the specification not stating the other necessary constraint: that the expression should be a valid expression when `x` is not const.

So, basically, I think fixing this incompatibility is fine, and the spec should be updated to make it clear why it's fine.

**Comment by gbracha** [comment]
I don't see the need to specify that the expression must be valid. That requirement exists independently.  The bug can be fixed in any case.

**Comment by lrhn** [comment]
I agree that the bug can and should be fixed.

I don't agree that the specification requires a potentially constant expression to be a valid expression when constructor parameters are not considered constant expressions. I can't find anything in the specification saying that `const C(x) : y = const D(x);` is not a valid const constructor. There is text saying that it is valid if  `const D(x)` is valid when `x` is considered a compile-time constant, and nothing else, so by that, it should be valid. No implementation actually allows it, so we have a spec/implementation disagreement. (or, alternatively, if the spec allows `const C(x): y = "$x";` but not `const C(x): y = const D(x);`, then there should be some text that distinguishes the two cases, and I can't find that - which is why I think it should be added).

**Comment by mhausner** [comment]
Committed patchset #1 (id:1) manually as
29e8118403f79cbe23a838776c1d132b53ca7ddb (presubmit successful).
